source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -77,30 +78,31 @@ class SimpleInsertContext : public InsertContext {
 
     body_.add(chunk);
     if (end_stream) {
-      commit();
+      ready_for_next_chunk(commit());
+    } else {
+      ready_for_next_chunk(true);
     }
-    ready_for_next_chunk(true);
   }
 
   void insertTrailers(const Http::ResponseTrailerMap& trailers,
                       InsertCallback insert_complete) override {
     ASSERT(!committed_);
     trailers_ = Http::createHeaderMap<Http::ResponseTrailerMapImpl>(trailers);
-    commit();
-    insert_complete(true);
+    insert_complete(commit());
   }
 
   void onDestroy() override {}
 
 private:
-  void commit() {
+  bool commit() {
     committed_ = true;
     if (VaryHeaderUtils::hasVary(*response_headers_)) {
-      cache_.varyInsert(key_, std::move(response_headers_), std::move(metadata_), body_.toString(),
-                        request_headers_, vary_allow_list_, std::move(trailers_));
+      return cache_.varyInsert(key_, std::move(response_headers_), std::move(metadata_),
+                               body_.toString(), request_headers_, vary_allow_list_,
+                               std::move(trailers_));
     } else {
-      cache_.insert(key_, std::move(response_headers_), std::move(metadata_), body_.toString(),
-                    std::move(trailers_));
+      return cache_.insert(key_, std::move(response_headers_), std::move(metadata_),
+                           body_.toString(), std::move(trailers_));
```

Comment:
this didn't really change this PR but should we be thinking about a zero-copy way to transfer the body?

Reply:
I think simple_http_cache is just supposed to be an example, not a serious implementation. But also I think zero-copy isn't really an option anyway because the original version is still being used for the response - you can't have one used for the response and a separate one for the cache with no copying. (And you can't use the same one because the response path is going to destroy it.)